File: x-baseproject/src/main/java/com/xia/baseproject/rxhttp/callback/AbstractFileCallback.java
package com.xia.baseproject.rxhttp.callback;

import android.arch.lifecycle.LifecycleOwner;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.blankj.utilcode.util.CloseUtils;
import com.blankj.utilcode.util.ThreadUtils;
import com.xia.baseproject.rxhttp.utils.FileUtils;
import com.xia.baseproject.rxhttp.utils.Platform;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;

import okhttp3.ResponseBody;

/**
 * @author xia
 * @date 2018/7/28.
 */
@SuppressWarnings("WeakerAccess")
public abstract class AbstractFileCallback extends Callback<File> {
    /**
     * 目标文件存储的文件夹路径
     */
    private String destFileDir;
    /**
     * 目标文件存储的文件名
     */
    private String destFileName;
    private DownloadTask mDownloadTask;

    public AbstractFileCallback(@NonNull LifecycleOwner lifecycleOwner, String destFileDir, String destFileName) {
        super(lifecycleOwner);
        this.destFileDir = destFileDir;
        this.destFileName = destFileName;
    }

    @SuppressWarnings("unchecked")
    @Override
    public File parseNetworkResponse(final ResponseBody responseBody) {
        if (mDownloadTask == null) {
            mDownloadTask = new DownloadTask(responseBody);
            ThreadUtils.executeByIo(mDownloadTask);
        }
        return null;
    }

    private class DownloadTask extends ThreadUtils.SimpleTask<File> {
        private ResponseBody mResponseBody;

        DownloadTask(ResponseBody responseBody) {
            mResponseBody = responseBody;
        }

        @Nullable
        @Override
        public File doInBackground() {
            return saveFile(mResponseBody);
        }

        @Override
        public void onSuccess(@Nullable File result) {
            onResponse(result);
        }
    }

    private File saveFile(ResponseBody responseBody) {
        final File file = FileUtils.createFile(destFileDir, destFileName);
        final InputStream is = responseBody.byteStream();
        BufferedInputStream bis = null;
        FileOutputStream fos = null;
        BufferedOutputStream bos = null;

        try {
            bis = new BufferedInputStream(is);
            fos = new FileOutputStream(file);
            bos = new BufferedOutputStream(fos);

            final long total = responseBody.contentLength();
            final byte[] data = new byte[1024 * 4];
            int count;
            long sum = 0;
            int lastProgress = 0;
            while ((count = bis.read(data)) != -1) {
                bos.write(data, 0, count);
                sum += count;
                final int progress = Math.round(sum * 100f / total);
                if (lastProgress != progress) {
                    Platform.post(() -> inProgress(progress, total));
                }
                lastProgress = progress;
            }
            bos.flush();
            fos.flush();
        } catch (IOException e) {
            onError(e.getMessage());
        } finally {
            CloseUtils.closeIO(responseBody, bos, fos, bis, is);
        }
        return file;
    }
}

File: fly-http/src/main/java/com/xia/fly/http/callback/AbstractFileCallback.kt
package com.xia.fly.http.callback

import androidx.lifecycle.LifecycleOwner
import com.blankj.utilcode.util.CloseUtils
import com.blankj.utilcode.util.ThreadUtils
import com.xia.fly.utils.FileUtils
import com.xia.fly.utils.Platform
import io.reactivex.functions.Action
import okhttp3.ResponseBody
import java.io.*

/**
 * @author xia
 * @date 2018/7/28.
 */
@Suppress("unused")
abstract class AbstractFileCallback(lifecycleOwner: LifecycleOwner,
                                    private val destFileDir: String,//目标文件存储的文件夹路径
                                    private val destFileName: String//目标文件存储的文件名
) : Callback<File>(lifecycleOwner) {

    private var mDownloadTask: DownloadTask? = null

    @Throws(Exception::class)
    override fun parseNetworkResponse(responseBody: ResponseBody) {
        if (mDownloadTask == null) {
            mDownloadTask = DownloadTask(responseBody)
            ThreadUtils.executeByIo(mDownloadTask)
        }
    }

    private inner class DownloadTask internal constructor
    (private val mResponseBody: ResponseBody) : ThreadUtils.SimpleTask<File>() {

        override fun doInBackground(): File? {
            return saveFile(mResponseBody)
        }

        override fun onSuccess(result: File?) {
            onResponse(result)
        }
    }

    private fun saveFile(responseBody: ResponseBody): File {
        val file = FileUtils.createFile(destFileDir, destFileName)
        val `is` = responseBody.byteStream()
        var bis: BufferedInputStream? = null
        var fos: FileOutputStream? = null
        var bos: BufferedOutputStream? = null

        try {
            bis = BufferedInputStream(`is`)
            fos = FileOutputStream(file)
            bos = BufferedOutputStream(fos)

            val fileSize = responseBody.contentLength()
            val data = ByteArray(1024 * 4)
            var read: Int
            var downloadSize: Long = 0
            var lastProgress = 0

            do {
                read = bis.read(data)
                if (read == -1) {
                    break
                }
                bos.write(data, 0, read)
                downloadSize += read.toLong()
                val progress = Math.round(downloadSize * 100f / fileSize)
                if (lastProgress != progress) {
                    Platform.post(Action { inProgress(progress.toFloat(), fileSize) })
                }
                lastProgress = progress
            } while (true)

            bos.flush()
            fos.flush()
        } catch (e: IOException) {
            onError(e)
        } finally {
            CloseUtils.closeIO(responseBody, bos, fos, bis, `is`)
        }
        return file
    }
}
